src/runtime/buffer.h

```diff
@@ -391,13 +405,16 @@ static inline u8 pop_u8(buffer b)
 
 static inline void push_u8(buffer b, u8 x)
 {
-    assert(buffer_extend(b, 1));
+    buffer_assert(buffer_extend(b, 1));
+    buffer_assert(b->start <= b->end);
+    buffer_assert(b->end <= b->length);
```

Comment:
This should be `<` rather than `<=` (or alternatively this assert can be moved to the end of the function, after `b->end++`)